Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Array)
+            {
+                throw new Exception("Json is not an array");
+            }
+            var jsonArray = (JArray)json;
+            var arrayType = MakeArrayType(type, elementType);
+            var result = (IList) Activator.CreateInstance(arrayType, jsonArray.Count);
+            var elementMetadata = fields[0];
+            for (var i = 0; i < jsonArray.Count; i++)
+            {
+                result[i] = ConvertToStructuredOrUnstructuredValue(elementType, elementMetadata, jsonArray[i]);
+            }
+            if (type != arrayType)
+            {
+                var listType = type.IsAbstract ? typeof(List<>).MakeGenericType(elementType) : type;
+                var list = (IList) Activator.CreateInstance(listType);
+                for (int i = 0; i < result.Count; i++)
+                {
+                    list.Add(result[i]);
+                }
+                return list;
+            }
+
+            return result;
+        }
+
+        private static object ConvertToMap(Type type, Type keyType, Type valueType, List<FieldMetadata> fields, JToken json)
+        {
+            if (keyType != typeof(string)
+                && keyType != typeof(int) && keyType != typeof(int?)
+                && keyType != typeof(long) && keyType != typeof(long?))
+            {
+                throw new Exception($"Unsupported key type of dictionary {keyType}");
+            }
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Object)
+            {
+                throw new Exception("Json is not an object");
+            }
+            if (fields == null || fields.Count != 2)
+            {
+                throw new Exception("Expected map to have 2 metadata fields");
+            }
+            var keyMetadata = fields[0];
+            var fieldMetadata = fields[1];
+            var jsonObject = (JObject)json;
+            var dictionaryType = type.IsAbstract ? typeof(Dictionary<,>).MakeGenericType(keyType, valueType) : type;
+            var result = (IDictionary) Activator.CreateInstance(dictionaryType);
+            using (var jsonEnumerator = jsonObject.GetEnumerator())
+            {
+                while (jsonEnumerator.MoveNext())
+                {
+                    var jsonPropertyWithValue = jsonEnumerator.Current;
+                    var fieldValue = jsonPropertyWithValue.Value;
+                    var key = IsTextMetadata(keyMetadata) || IsFixedMetadata(keyMetadata)
+                        ? ConvertToUnstructuredType(keyMetadata, keyType, jsonPropertyWithValue.Key)
+                        : throw new Exception("Could not recognize key type for map");
```

Comment:
It's rather not supported - I believe what happens is that server should not send structured object as a key in this conversion, correct?

Reply:
On the sever side the key of map can be only TEXT or INTEGER. So this exception could be only thrown if on the sever side new types became possible to be used as a key in a map, but we did not update the driver. I changed to: $"Unsupported key type for map {keyMetadata.type}"